File: solutions/ocaml/prime-factors/1/prime_factors.ml
let rec find_div (n : int64) (p : int64) : int64 = 
    if Int64.rem n p = 0L then p else find_div n (Int64.add p 1L)
let rec factors_of n = 
    if n = 1L then [] else
        let p = find_div n 2L in
        p :: factors_of (Int64.div n p)

File: solutions/fsharp/prime-factors/1/PrimeFactors.fs
module PrimeFactors

let rec fs (n : int64) (p : int64) (r : int list) : int list =
    match n with
    | 1L -> r
    | _ -> 
        if n % p = 0 then 
            fs (n / p) p (List.append r [int p]) 
        else
            if n / p < p then 
                fs n n r 
            else 
                fs n (p + 1L) r

let factors number = fs number 2 []